### PR TITLE
Batch mode for frontend subsystems 1/N

### PR DESCRIPTION
**Pull request by graydon**
A bunch of work following @davidungar's work but further downstream in the compiler, altering the various subsystems called by FrontendTool (Sema/Serialization/SIL) to accept and work with multiple primary files. Theoretically NFC but see below.

You can't actually _use_ this without disabling a couple assertions upstream that the user only provided one primary, but that's an intentional part of @davidungar's plan :)

Passes swift tests but needs to be accompanied by a matching LLDB change.

There are a couple bits I'd like to see scrutinized in review here, each marked with a comment. The change to lib/SILGen/SILGenExpr.cpp is I _think_ NFC but I'm not 100% sure, and the FIXME-marked bit on line 793 of lib/FrontendTool/FrontendTool.cpp suggests we might have drifted to the point of having redundant logic that can be eliminated.

### CODE REVIEW

**Review by davidungar — COMMENTED** [review]
"notePrimarySourceFile" and its ilk could be more suggestive of multiple primaries. "noteAPrimarySourceFile"? or "noteAnotherPrimarySourceFile"? or "addThisFileToTheSetOfPrimaries"?

**Review by davidungar — APPROVED** [review]
The parts of this I know about--not all that much--look fine to me! Good stuff.

File: include/swift/Frontend/Frontend.h [hunk 1]
```diff
@@ -336,14 +337,31 @@ class CompilerInstance {
...
+  llvm::SetVector<SourceFile*> PrimarySourceFiles;
+
+  /// Return whether there is an entry in PrimaryInputs for buffer \BufID.
+  bool bufferIDIsAPrimaryInput(unsigned BufID) const {
+    return PrimaryBufferIDs.count(BufID) != 0;
+  }
+
+  /// Record in PrimaryBufferIDs the fact that \BufID is a primary.
+  /// If \BufID is already in the set, do nothing.
+  void notePrimaryInputBuffer(unsigned BufID);
```
Comment (jrose-apple) [inline include/swift/Frontend/Frontend.h]: Naming nitpicks: "record" or "add" feels stronger than "note".

File: include/swift/Frontend/Frontend.h [hunk 2]
```diff
@@ -336,14 +337,31 @@ class CompilerInstance {
...
+  llvm::SetVector<unsigned> PrimaryBufferIDs;
 
-  SourceFile *PrimarySourceFile = nullptr;
+  /// Identifies the set of SourceFiles that are considered primaries. An
+  /// invariant is that any SourceFile in this set with an associated
+  /// buffer will also have its buffer ID in PrimaryBufferIDs.
+  llvm::SetVector<SourceFile*> PrimarySourceFiles;
+
+  /// Return whether there is an entry in PrimaryInputs for buffer \BufID.
+  bool bufferIDIsAPrimaryInput(unsigned BufID) const {
```
Comment (jrose-apple) [inline include/swift/Frontend/Frontend.h]: Naming nitpicks: pulling the "is" up front makes it more obvious at the call site that this is a predicate: `isBufferIDAPrimaryInput`, `isBufferPrimaryInput` or even just `isPrimaryInput`.

File: include/swift/Subsystems.h [hunk 3]
```diff
@@ -242,10 +243,10 @@ namespace swift {
   performSILGeneration(FileUnit &SF, SILOptions &options,
                        Optional<unsigned> StartElem = None);
 
-  using ModuleOrSourceFile = PointerUnion<ModuleDecl *, SourceFile *>;
-
-  /// Serializes a module or single source file to the given output file.
-  void serialize(ModuleOrSourceFile DC, const SerializationOptions &options,
+  /// Serializes a module and zero or more source files to the given output file.
```
Comment (jrose-apple) [inline include/swift/Subsystems.h]: This isn't right. We still need one swiftmodule per input file, or incremential compilation won't work.

**Review by jrose-apple — CHANGES_REQUESTED** [review]
This is the wrong way to go. In order to keep supporting incremental builds, we need to keep emitting N swiftmodules, object files, and other auxiliaries when given N primary files as input.

**Comment by graydon** [comment]
Ah, crumpets. Well good to find out early I guess! @jrose-apple do you think the non-Serialization parts are usable? I.e. the bits related to Sema and SILGen/SILOpt? Or do we need a separate AST ModuleDecl and SILModule per primary as well, if we're going to be emitting a separate .swiftmodule file per primary?

**Comment by jrose-apple** [comment]
I think our goal was to have a combined AST ModuleDecl, but separate SILModules. That way we still save repeated type-checking work. The tricky bit was dealing with SILGen for "external declarations" i.e. "ClangImporter things that need SILGen"; for now the easiest thing is probably to just SILGen them all into every SIL module.

**Comment by jrose-apple** [comment]
The frontendy things (the first and third commits) seem pretty much fine. I didn't take a close look at the second or fourth commits.

**Comment by graydon** [comment]
K, will do. Sorry!

**Comment by graydon** [comment]
Closing this out and opening new one with simpler / corrected multiplicity approach.

**Pull request closed without merging**